### PR TITLE
CSP-1571 Sending Employer-led email notifications (to Providers)

### PR DESCRIPTION
**Pull request by shaun-downey-education**
- Added a new function (**SendEmployerLedNotificationsFunction**) to handle the sending of notifications for employer led email notifications (Providers).
- Added new unit test project **SFA.DAS.PR.Data.UnitTests**.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -83,6 +83,7 @@ local.settings.json file
     "EnvironmentName": "LOCAL",
     "AzureWebJobsServiceBus": "",
     "UpdateProvidersFunctionSchedule": "0 0/3 * * * *",
+    "SendEmployerLedNotificationsFunctionSchedule": "*/5 * * * *",
     "AzureWebJobs.UpdateProvidersFunction.Disabled": "true"
```
Comment (DevenKShah) [inline README.md]: Add your functions disable flag here ` "AzureWebJobs.SendNotificationsFunction.Disabled": "true"`

File: azure/template.json [hunk 2]
```diff
@@ -267,8 +268,12 @@
...
                                 },
                                 {
-                                    "name":  "UpdateProvidersFunctionSchedule",
-                                    "value": "[variables('updateProvidersFunctionSchedule')]"
+                                  "name": "UpdateProvidersFunctionSchedule",
+                                  "value": "[variables('updateProvidersFunctionSchedule')]"
+                                },
+                                {
+                                  "name": "SendNotificationsFunctionSchedule",
+                                  "value": "[variables('sendNotificationsFunctionSchedule')]"
```
Comment (DevenKShah) [inline azure/template.json]: Align this to match rest of the file

File: src/SFA.DAS.PR.Data/Repositories/NotificationRepository.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+public interface INotificationRepository
+{
+    Task<List<Notification>> GetPendingNotifications(int batchSize, NotificationType notificationType, CancellationToken cancellationToken);
+}
+
+public class NotificationRepository(IProviderRelationshipsDataContext _providerRelationshipsDataContext) : INotificationRepository
+{
+    public async Task<List<Notification>> GetPendingNotifications(int batchSize, NotificationType notificationType, CancellationToken cancellationToken)
+    {
+        return await _providerRelationshipsDataContext.Notifications
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Data/Repositories/NotificationRepository.cs:16]: The query should retrieve data in order of creation (FIFO), add `OrderBy(n > n.CreatedDate)`

File: src/SFA.DAS.PR.Jobs.UnitTests/SFA.DAS.PR.Jobs.UnitTests.csproj [hunk 4]
```diff
@@ -29,9 +29,17 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
     </PackageReference>
     <PackageReference Include="SFA.DAS.Testing.AutoFixture" Version="17.1.94" />
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs.UnitTests/SFA.DAS.PR.Jobs.UnitTests.csproj:31]: We are not using xunit, so why this?

File: src/SFA.DAS.PR.Jobs/Configuration/TemplateConfiguration.cs
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Configuration/TemplateConfiguration.cs:1]: Suggest you rename the type to `NotificationTemplateConfiguration`

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        _logger.LogInformation("{FunctionName} has been triggered.", nameof(SendNotificationsFunction));
+
+        List<Notification> notifications = await _notificationRepository.GetPendingNotifications(
+            _notificationsConfiguration.BatchSize, 
+            NotificationType.Provider, 
+            cancellationToken
+        );
+
+        if(notifications.Count == 0)
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs]: There is no need to return anything, is there? I would rather just log the count anyway, so we can see the job is working.

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        _providerRelationshipsDataContext = providerRelationshipsDataContext;
+        _notificationRepository = notificationRepository;
+        _tokenService = tokenService;
+
+        _notificationsConfiguration = new NotificationsConfiguration();
+        configuration.GetSection("ApplicationConfiguration:Notifications").Bind(_notificationsConfiguration);
+    }
+
+    [Function(nameof(SendNotificationsFunction))]
+    public async Task<int> Run([TimerTrigger("%SendNotificationsFunctionSchedule%", RunOnStartup = true)] TimerInfo timer, FunctionContext executionContext, CancellationToken cancellationToken)
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs]: Set runonstartup to false

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        if(notifications.Count == 0)
+        {
+            return 0;
+        }
+
+        int processedCount = 0;
+
+        foreach (Notification notification in notifications)
+        {
+            processedCount += await ProcessNotification(notification, executionContext, cancellationToken);
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs]: I think we need to add parallelism here else the job will take too long to finish. Try adding 500 notifications locally and see how it behaves.

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private async Task<SendEmailCommand> CreateSendEmailCommand(Notification notification, CancellationToken cancellationToken)
+    {
+        TemplateConfiguration? templateConfiguration = _notificationsConfiguration.NotificationTemplates.Find(a => a.TemplateName == notification.TemplateName);
+
+        if(templateConfiguration == null || string.IsNullOrWhiteSpace(templateConfiguration.TemplateId))
+        {
+            throw new ArgumentNullException($"Template {notification.TemplateName} does not exist in the configuration for notification {notification.Id}.");
+        }
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs]: This should never happen and hence we should not code around it. If it happens, then it will naturally fail and we should let that exception bubble up.

File: src/SFA.DAS.PR.Jobs/Configuration/TemplateConfiguration.cs
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Jobs/Configuration/TemplateConfiguration.cs:1]: The object is the parent for both the batch size and template list so would that make sense?

File: azure/template.json [hunk 9]
```diff
@@ -1,345 +1,352 @@
...
+  },
+  "variables": {
+    "deploymentUrlBase": "https://raw.githubusercontent.com/SkillsFundingAgency/das-platform-building-blocks/master/templates/",
+    "resourceNamePrefix": "[toLower(concat('das-', parameters('resourceEnvironmentName'),'-', parameters('serviceName')))]",
+    "resourceGroupName": "[concat(variables('resourceNamePrefix'), '-rg')]",
+    "functionAppName": "[concat(variables('resourceNamePrefix'), '-fa')]",
+    "appServicePlanName": "[concat(variables('resourceNamePrefix'), '-asp')]",
+    "configNames": "SFA.DAS.PR.Jobs",
+    "updateProvidersFunctionSchedule": "0 0 0 * * *",
+    "sendNotificationsFunctionSchedule": "*/5 * * * *"
```
Comment (DevenKShah) [inline azure/template.json]: azure functions takes 6 arguments in cron expression `0 */5 * * * *`. There are examples [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-timer?tabs=python-v2%2Cisolated-process%2Cnodejs-v4&pivots=programming-language-csharp#ncrontab-examples)

File: README.md [hunk 10]
```diff
@@ -83,7 +83,9 @@ local.settings.json file
     "EnvironmentName": "LOCAL",
     "AzureWebJobsServiceBus": "",
     "UpdateProvidersFunctionSchedule": "0 0/3 * * * *",
-    "AzureWebJobs.UpdateProvidersFunction.Disabled": "true"
+    "SendNotificationsFunctionSchedule": "*/5 * * * *",
```
Comment (DevenKShah) [inline README.md]: cron expression to start with `0 */5 * * * *`

File: src/SFA.DAS.PR.Jobs/Extensions/AddServiceRegistrationsExtension.cs [hunk 11]
```diff
@@ -12,12 +12,41 @@ public static class AddServiceRegistrationsExtension
...
+        
         services
-            .AddHttpClient()
-            .AddRefitClient<IRoatpServiceApiClient>()
-            .ConfigureHttpClient(c => c.BaseAddress = new Uri(roatpServiceApiConfiguration.Url))
-            .AddHttpMessageHandler(() => new InnerApiAuthenticationHeaderHandler(new AzureClientCredentialHelper(), roatpServiceApiConfiguration.Identifier));
+            .RegisterServices()
+            .RegisterRoatpServiceApiClient(configuration)
+            .RegisterPasAccountApiClient(configuration)
+            .AddHttpClient();
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Extensions/AddServiceRegistrationsExtension.cs]: Move AddHttpClient above api client registrations.

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        _logger = logger;
+        _functionEndpoint = functionEndpoint;
+        _providerRelationshipsDataContext = providerRelationshipsDataContext;
+        _notificationRepository = notificationRepository;
+        _tokenService = tokenService;
+        _pasAccountApiClient = pasAccountApiClient;
+
+        _notificationsConfiguration = new NotificationsConfiguration();
+        configuration.GetSection("ApplicationConfiguration:Notifications").Bind(_notificationsConfiguration);
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs]: Since the function is running every 5 minutes, I would use IOptions pattern to register ApplicationConfiguration and have dependency on IOptions here so we are not reading the configuration object and deserialising as many times the function runs.

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+                processedCount += await ProcessProviderNotification(notification, executionContext, cancellationToken);
+            }
+
+            await _providerRelationshipsDataContext.SaveChangesAsync(cancellationToken);
+        }
+
+        _logger.LogInformation("{FunctionName} - Processed {ProcessedCount} notifications.", nameof(SendNotificationsFunction), processedCount);
+    }
+
+    private async Task<int> ProcessProviderNotification(Notification notification, FunctionContext executionContext, CancellationToken cancellationToken)
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs]: I would just rename this method by adding prefix `TryProcess...` so we know that this method can fail and failure is handled.

File: src/SFA.DAS.PR.Jobs/Infrastructure/IRoatpServiceApiClient.cs [hunk 14]
```diff
@@ -5,6 +5,6 @@ namespace SFA.DAS.PR.Jobs.Infrastructure;
 
 public interface IRoatpServiceApiClient
 {
-    [Get("/api/v1/fat-data-export")]
+    [Post("/api/v1/fat-data-export")]
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Infrastructure/IRoatpServiceApiClient.cs]: Why was this changed to POST it should be GET

File: src/SFA.DAS.PR.Jobs/Services/TokenService.cs
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Services/TokenService.cs:1]: Can we rename this to say `NotificationTokenService` it can easily get mixed up for auth tokens.

File: src/SFA.DAS.PR.Jobs/Infrastructure/IRoatpServiceApiClient.cs (hunk 14 above)
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Jobs/Infrastructure/IRoatpServiceApiClient.cs]: @DevenKShah Not sure this was intentional, i'll revert it.

File: src/SFA.DAS.PR.Jobs.UnitTests/SFA.DAS.PR.Jobs.UnitTests.csproj (hunk 4 above)
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Jobs.UnitTests/SFA.DAS.PR.Jobs.UnitTests.csproj:31]: **SFA.DAS.Testing.AutoFixture** is used to use the **CreateMany** extension on Fixture

var providers = fixture
    .Build<Provider>()
    .CreateMany(p => p.Ukprn, ValidUkprns)
    .ToList();

Comment (DevenKShah) [reply src/SFA.DAS.PR.Jobs.UnitTests/SFA.DAS.PR.Jobs.UnitTests.csproj:31]: Sorry I think it may have moved or removed, but there was a reference to  XUnit

Comment (DevenKShah) [reply src/SFA.DAS.PR.Jobs.UnitTests/SFA.DAS.PR.Jobs.UnitTests.csproj:31]: I can see `<PackageReference Include="xunit.extensibility.core" Version="2.9.0" />` is still referenced

Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Jobs.UnitTests/SFA.DAS.PR.Jobs.UnitTests.csproj:31]: @DevenKShah I've removed that package reference there now.

**Review by DevenKShah — DISMISSED** [review]

**Review by DevenKShah — DISMISSED** [review]

File: src/SFA.DAS.PR.Jobs/Services/NotificationTokenService.cs [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        {
+            case NotificationType.Provider:
+                {
+                    Provider? provider = await _providersRepository.GetProvider(notification.Ukprn!.Value, cancellationToken);
+
+                    AccountLegalEntity? accountLegalEntity = await _accountLegalEntityRepository.GetAccountLegalEntity(notification.AccountLegalEntityId!.Value, cancellationToken);
+
+                    emailTokens = new()
+                    {
+                        { EmailTokens.ProviderPortalUrlToken, _notificationConfigurationOptions.Value.ProviderPortalUrl},
```
Comment (alandburns) [inline src/SFA.DAS.PR.Jobs/Services/NotificationTokenService.cs:36]: The Email template merge field is called provideraccountsweb

**Review by debrr07 — APPROVED** [review]

**Review by MarkFCain — APPROVED** [review]

File: src/SFA.DAS.PR.Jobs/Services/NotificationTokenService.cs (hunk 15 above)
Comment (DevenKShah) [reply src/SFA.DAS.PR.Jobs/Services/NotificationTokenService.cs:36]: This is a constant named based on the config setting, so we know how it maps. The value of the constant matches the email template.
